### PR TITLE
SAK-46974 MSGCNTR Provide private message scheduling

### PR DESCRIPTION
**Pull request by victorGomollon**
[SAK-46974: Messages. Provide private message scheduling](https://sakaiproject.atlassian.net/browse/SAK-46974)

### CODE REVIEW

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/ui/PrivateMessageManagerImpl.java [hunk 1]
```diff
@@ -1868,6 +1905,19 @@ public boolean isAllowToFieldGroups(User user)
       return false;
   }
 
+  @Override
+  public boolean isAllowToFieldGroups(User user, String contextId)
+  {
+    if (log.isDebugEnabled())
```
Comment (ern) [inline msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/ui/PrivateMessageManagerImpl.java:1911]: use the new log message style

File: msgcntr/messageforums-hbm/src/java/org/sakaiproject/component/app/messageforums/dao/hibernate/MessageImpl.java [hunk 2]
```diff
@@ -479,6 +481,23 @@ public void setNumReaders(Integer numReaders) {
 
 	}
 
+	public Boolean getScheduler()
```
Comment (ern) [inline msgcntr/messageforums-hbm/src/java/org/sakaiproject/component/app/messageforums/dao/hibernate/MessageImpl.java:484]: lombok'd

File: msgcntr/messageforums-api/src/bundle/org/sakaiproject/api/app/messagecenter/bundle/Messages.properties [hunk 3]
```diff
@@ -685,6 +688,10 @@ pvt_replied=Already replied
 
 pvt_no_subject=No subject
 
+#SAK-46974:
+pvt_scheduler_send= Program message sending
+pvt_scheduler_send_as_email= Program the message to be sent at the selected date and time
```
Comment (fostersdesign) [inline msgcntr/messageforums-api/src/bundle/org/sakaiproject/api/app/messagecenter/bundle/Messages.properties:693]: How about something like:

```suggestion
pvt_scheduler_send=Schedule when this message will be sent
pvt_scheduler_send_as_email=Schedule the message to be sent at the selected date and time
```

**Comment by fostersdesign** [comment]
Is the scheduled date validated to make sure it is in the future?

**Comment by victorGomollon** [comment]
> Is the scheduled date validated to make sure it is in the future?

Yes

**Comment by SedueRey** [comment]
@ern We are trying to figure out how to test the service properly, since it runs on a separate thread and we don't quite know how to do it. What we had tested on Springboot does not work on sakai.

Any ideas you can give us? 

Thanks!!!

File: msgcntr/messageforums-api/src/bundle/org/sakaiproject/api/app/messagecenter/bundle/Messages.properties (hunk 3 above)
Comment (jesusmmp) [reply msgcntr/messageforums-api/src/bundle/org/sakaiproject/api/app/messagecenter/bundle/Messages.properties:693]: @victorGomollon change this pls

**Review by juanjmerono — APPROVED** [review]

File: msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/PrivateMessagesTool.java [hunk 4]
```diff
@@ -2706,9 +2805,29 @@ private PrivateMessage getPvtMsgForward(PrivateMessage currentMessage, boolean i
...
+		    }
+	    }
+	    rrepMsg.setScheduler(booleanSchedulerSend);
+
+	    Map<User, Boolean> recipients = getRecipients();
+
+	    if (booleanSchedulerSend && !rrepMsg.getDraft()) {
+		    rrepMsg.setScheduledDate(openDate);
+		    schedulerMessage(rrepMsg, isSendEmail());
+		    resetComposeContents();
```
Comment (stetsche) [inline msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/PrivateMessagesTool.java:2822]: I was wondering if there is a missing return here, as the scheduled invocation will also create an event.

File: msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/PrivateMessagesTool.java [hunk 5]
```diff
@@ -2973,7 +3099,27 @@ private PrivateMessage processPvtMsgReplyAllSendHelper(boolean preview, Boolean
...
+	            setErrorMessage(getResourceBundleString(DATE_PGR_MSG_ERROR));
+	            return null;
+	        }
+		  }
+		    rrepMsg.setScheduler(booleanSchedulerSend);
+
+	        if(booleanSchedulerSend && !rrepMsg.getDraft()) {
+	            rrepMsg.setScheduledDate(openDate);
+	            schedulerMessage(rrepMsg, isSendEmail());
+	            resetComposeContents();
```
Comment (stetsche) [inline msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/PrivateMessagesTool.java:3114]: This is the other place where I feel like a return could be missing.

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/scheduler/PrivateMessageSchedulerServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,672 @@
...
+
+		Map<User, Boolean> recipients = getRecipients(pvtMsg);
+
+		pvtMsg.setCreated(pvtMsg.getScheduledDate());
+		pvtMsg.setModified(pvtMsg.getScheduledDate());
+		pvtMsg.setScheduledDate(null);
+		pvtMsg.setScheduler(false);
+		pvtMsg.setDraft(false);
+
+		prtMsgManager.sendPrivateMessage(pvtMsg, recipients, false);
```
Comment (ottenhoff) [inline msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/scheduler/PrivateMessageSchedulerServiceImpl.java:122]: why is the sendingEmail always false? what if the user selected to send an email in the future?

Comment (SedueRey) [reply msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/scheduler/PrivateMessageSchedulerServiceImpl.java:122]: I don't remember right now why it was marked as false and Victor is currently on paternity leave, he'll be back in a month or so, so any ideas you have to change this behaviour are valid, have you had any bugs with this?

Comment (ottenhoff) [reply msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/scheduler/PrivateMessageSchedulerServiceImpl.java:122]: @jesusmmp picked up this bug. thanks to Jesus. but the bug is clear: no email is sent even when the instructor wants and expects email to be sent

**Pull request merged**